### PR TITLE
DnD constructors

### PR DESCRIPTION
**Pull request by nickbattam-tessella**
Create constructors for remaining dimensions (0d, 1d, 3d, 4d)
This has moved virtually all the code out of the concrete classes.

### CODE REVIEW

File: horace_core/sqw/@DnDBase/parse_args_.m [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+if isa(input, 'SQWDnDBase')
+    if isa(input, class(obj))
+        args.dnd_obj = input;
+    elseif isa(input, 'sqw')
+        args.sqw_obj = input;
+    else
+        error([upper(class(obj)), ':' class(obj)], ...
+            [upper(class(obj)) ' cannot be constructed from an instance of this object "' class(input) '"']);
+    end
```
Comment (cmarooney-stfc) [inline horace_core/sqw/@DnDBase/parse_args_.m:26]: did you mean "instance of this class"?

File: horace_core/sqw/@DnDBase/parse_args_.m [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+elseif is_string(parser.Results.input)
+    args.filename = input;
```
Comment (cmarooney-stfc) [inline horace_core/sqw/@DnDBase/parse_args_.m:28]: as you have unpacked parser.Results.input as input, may be better to keep the shorter name here

**Review by cmarooney-stfc — CHANGES_REQUESTED** [review]

File: _test/test_dnd_class/test_dnd_constructor.m [hunk 3]
```diff
@@ -80,13 +106,20 @@ function test_d2d_get_returns_set_properties(obj)
...
         assertEqualToTol(dnd_copy, dnd_obj);
-    end
+  end
+    function test_copy_constructor_clones_d4d_object(obj)
+          dnd_obj = d4d();
+          dnd_copy = d4d(dnd_obj);
+
+          assertTrue(isa(dnd_obj, 'd4d'));
+          assertEqualToTol(dnd_copy, dnd_obj);
+      end
```
[the remark this answers is not in the document]
Comment (nickbattam-tessella) [reply _test/test_dnd_class/test_dnd_constructor.m:122]: The idea is to check that changes made to one object are not visible in the second.

The `test_copy_constructor_clones_...` tests confirm that after a copy the new new objects are "equal" (as far as the `equal_to_tol` definition goes.
This test builds on that and verifies that after that copy I can change some attributes and get an object that is no longer equal.

File: _test/test_dnd_class/test_dnd_constructor.m [hunk 4]
```diff
@@ -26,20 +26,46 @@
...
+        assertEqual(numel(dnd_obj.pax), 3);
+        assertEqual(dnd_obj.dimensions(), 3);
     end
+    function test_d4d_constructor_returns_4d_instance(obj)
+        dnd_obj = d4d();
+
+        assertEqual(numel(dnd_obj.pax), 4);
+        assertEqual(dnd_obj.dimensions(), 4);
+    end
+
```
[the remark this answers is not in the document]
Comment (nickbattam-tessella) [reply _test/test_dnd_class/test_dnd_constructor.m:68]: It is also one pixel long.

File: horace_core/sqw/@DnDBase/parse_args_.m (hunk 2 above)
Comment (hsautessella) [reply horace_core/sqw/@DnDBase/parse_args_.m:28]: Also, `input` is a Matlab built-in, so using a different name would probably be best.

File: _test/test_dnd_class/test_dnd_constructor.m [hunk 5]
```diff
@@ -65,8 +121,27 @@ function test_d2d_contains_expected_properties(obj)
         end
     end
 
+    function test_d0d_get_returns_set_properties(obj)
+        dnd_obj = d0d();
+        obj.assert_dnd_get_returns_set_properties(dnd_obj);
+    end
+    function test_d1d_get_returns_set_properties(obj)
```
Comment (hsautessella) [inline _test/test_dnd_class/test_dnd_constructor.m:128]: Is the intent of not having whitespace between these tests to group them? Would a comment, or even a section delimiter (`%%`), be a better way of grouping them? If we ever decide on a formatter this grouping will be lost

**Review by hsautessella — CHANGES_REQUESTED** [review]
Just a couple of small things

**Pull request merged**